framework/File.cpp

```diff
@@ -33,6 +33,11 @@ If you have questions concerning this license or the applicable additional terms
 
 #define	MAX_PRINT_MSG		4096
 
+#if (__GNUC__ >= 5) || (__GNUC__ == 4 && __GNUC_MINOR__ >= 6)
```

Comment:
should be like `#if defined(__GNUC__) && ((__GNUC__ >= 5) || (__GNUC__ == 4 && __GNUC_MINOR__ >= 6))`
otherwise there will be warning with other compilers that you're using an unknown symbol (`__GNUC__` which is only known to GCC and compatibles like clang) in the `#if`

Other problem with this: While clang *does* define `__GNUC__`, they pretend to be GCC 4.2? They still have `-Wformat-security` and *do* support `#pragma GCC diagnostic push` etc, so it *might* be necessary to handle that as well.
Or just stick to `#ifdef __GNUC__`, if someone *really* still uses a GCC version older than 4.6, it will most probably just show a warning about the unknown pragma and carry on, at least GCC 4.1 does: https://gcc.godbolt.org/z/cY7Wqqsah

Reply:
No doubt I'll change my mind about this at some point, but I think it's useful to try to define version ranges correctly, so I'll go with the `#if defined(...)` suggestion to begin with.

As for `clang`.. eh.  It probably wasn't intentional, but presenting a version declaration without being compatible with it seems likely to cause exactly the kind of problem described.  I'll read the link and some more about that all soon.

(agreed it's probably unlikely that this would be compiled on older GCC versions, but who knows)

Reply:
it *is* compatible with GCC 4.2, but it's also (partially) compatible with newer versions of GCC

and to put it bluntly, clang is *way* more relevant than 12+ years old GCC releases..

anyway, you could explicitly identify clang with `#if defined(__clang__)`

Reply:
Kind of a weird tangential story, but: not too long ago I was working on an extremely small library (less than 300 lines of code), but iterating rapidly on releases (major and minor increments, depending on the compatibility of changes).

Fairly late-on during development I encountered what I considered to be a fairly showstopper bug, and realized that it affected almost _all_ of the release history.  I fixed it in a forward-looking way first of all, but then I went back to backport it to _every_ prior affected point release.

While doing that I realized: hey, some of these previous releases are actually in some ways better than the most recent one.  They have some quirks, but performance, or compatibility, or readability or so on might be better.

After that I made a bit more of an effort to backport improvements from modern versions back into previous versions, hoping to improve all of the timelines.

It's probably not relevant for GCC/clang, and in fact, it's probably a completely irrelevant story, but I figured I'd share it anway.